File: boletim-ocorrencia/src/main/java/br/edu/utfpr/td/tsi/sistema/boletim/ocorrencia/business/ReportManager.java
package br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.business;

import br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.dao.InMemoryReportDAO;
import br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.model.StolenVehicleReport;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.DateTimeException;
import java.time.LocalDate;
import java.util.List;

@Component
public class ReportManager implements ReportRules {
    @Autowired
    private InMemoryReportDAO reportDAO;

    @Autowired
    private PartManager partManager;

    @Override
    public void register(StolenVehicleReport report) {
        if (isDataValid(report.getOccurrenceDate())) {
            report.setPart(partManager.hidePartInformations(report.getPart()));
            reportDAO.register(report);
        } else {
            throw new DateTimeException("A data informada é maior que a data atual!");
        }
    }

    @Override
    public void update(StolenVehicleReport report) {
        reportDAO.update(report);
    }

    @Override
    public List<StolenVehicleReport> readAll() {
        return reportDAO.readAll();
    }

    @Override
    public StolenVehicleReport readById(int id) {
        return reportDAO.readById(id);
    }

    @Override
    public List<StolenVehicleReport> readByCity(String city) {
        return reportDAO.readByCity(city);
    }

    @Override
    public List<StolenVehicleReport> readByDayTime(String dayTime) {
        return reportDAO.readByDayTime(dayTime);
    }

    @Override
    public void remove(int id) {
        reportDAO.remove(id);
    }


    public boolean isDataValid(LocalDate reportDate) {
        LocalDate currentDate = LocalDate.now();
        return reportDate.isBefore(currentDate) || reportDate.isEqual(currentDate);
    }
}

File: boletim_python_v/occurance_report_system/occurance_report_system/occurance_report/busisness_rules/ReportManager.py
from datetime import datetime
from ..dao import InMemoryReportDao
from ..models import stolen_vehicle_report


class ReportManager:

    def __init__(self):
        self.report_dao = None

    def init(self):
        self.report_dao = in_memory_report_dao

    def register(self, report):
        if self.is_data_valid(report.occurrence_date):
            self.report_dao.register(report)
        else:
            raise ValueError("A data informada é maior que a data atual!")

    def read_all(self):
        return self.report_dao.read_all()

    def read_by_id(self, id):
        return self.report_dao.read_by_id(id)

    def read_by_city(self, city):
        return self.report_dao.read_by_city(city)

    def read_by_day_time(self, day_time):
        return self.report_dao.read_by_day_time(day_time)

    def remove(self, id):
        self.report_dao.remove(id)

    def is_data_valid(self, report_date):
        current_date = datetime.now().date()
        return report_date.date() <= current_date
